### PR TITLE
HIVE-26551: Support CREATE TABLE LIKE FILE for ORC

### PR DESCRIPTION
**Pull request by zhangbutao**
### Why are the changes needed?

CREATE TABLE LIKE syntax was added in https://github.com/apache/hive/pull/3443, and this PR added the ability to create table   based on existing orc files.
It is convenient for user to create a table based on a orc file.

### Does this PR introduce _any_ user-facing change?

NO

### How was this patch tested?

mvn test -Dtest=TestMiniLlapLocalCliDriver -Dqfile=create_table_like_file_orc.q -Dtest.output.overwrite=true

### CODE REVIEW

**Comment by jfsii** [comment]
Thanks for doing this work @zhangbutao ! Very happy to see ORC support. I'll take a look and give this a try on my machine.

File: ql/src/java/org/apache/hadoop/hive/ql/io/orc/OrcSerde.java [hunk 1]
```diff
@@ -117,4 +130,87 @@ public ObjectInspector getObjectInspector() throws SerDeException {
...
+    return buffer.toString();
+  }
+
+  static void getStructFieldName(StringBuilder buffer, String name) {
+    if (UNQUOTED_NAMES.matcher(name).matches()) {
+      buffer.append(name);
+    } else {
+      buffer.append('`');
+      buffer.append(name.replace("`", "``"));
+      buffer.append('`');
```
Comment (ayushtkn) [inline ql/src/java/org/apache/hadoop/hive/ql/io/orc/OrcSerde.java:213]: nit: Concurrent StringBuilder Can be reused.
```
      buffer.append('`').append(name.replace("`", "``")).append('`');
```

File: ql/src/java/org/apache/hadoop/hive/ql/io/orc/OrcSerde.java [hunk 2]
```diff
@@ -117,4 +130,87 @@ public ObjectInspector getObjectInspector() throws SerDeException {
...
+      break;
+    default:
+      throw new IllegalArgumentException("ORC doesn't handle " +
+          fieldType.getCategory());
+    }
+    return buffer.toString();
+  }
+
+  static void getStructFieldName(StringBuilder buffer, String name) {
+    if (UNQUOTED_NAMES.matcher(name).matches()) {
```
Comment (ayushtkn) [inline ql/src/java/org/apache/hadoop/hive/ql/io/orc/OrcSerde.java:208]: Is the intent of this just to ensure there aren't any quoted `name`, I am not sure, but do explore 
```
!name.startsWith("'") && !name.endsWith("'")
```
Which lands up being cheaper.

File: common/src/java/org/apache/hadoop/hive/ql/ErrorMsg.java [hunk 3]
```diff
@@ -524,6 +524,7 @@ public enum ErrorMsg {
   CTLF_MISSING_STORAGE_FORMAT_DESCRIPTOR(20021, "Failed to find StorageFormatDescriptor for file format ''{0}''", true),
   PARQUET_FOOTER_ERROR(20022, "Failed to read parquet footer:"),
   PARQUET_UNHANDLED_TYPE(20023, "Unhandled type {0}", true),
+  ORC_FOOTER_ERROR(20024, "Failed to read orc footer:"),
```
Comment (ayushtkn) [inline common/src/java/org/apache/hadoop/hive/ql/ErrorMsg.java:527]: Well the colon is there in ``PARQUET_FOOTER_ERROR`` but I am not sure why it is so, when this exception is thrown, post the colon it is the trace, should have been a period according to me, but let it be to be in sync with PARQUET_FOOTER_ERROR. But just out of curiosity if you know the reason behind it do let me know

**Review by ayushtkn — COMMENTED** [review]
Had a look, and things look good to me, the test covers kind of a lot of things.
Lets wait for @jfsii to have a look

File: ql/src/java/org/apache/hadoop/hive/ql/io/orc/OrcSerde.java (hunk 2 above)
Comment (zhangbutao) [reply ql/src/java/org/apache/hadoop/hive/ql/io/orc/OrcSerde.java:208]: Actually, this code snippet was borrowed from ORC repo.
https://github.com/apache/orc/blob/6a74eef74b8101c7396f196f1a27f5e124a005f6/java/core/src/java/org/apache/orc/TypeDescription.java#L667-L673

And related ORC jira is https://issues.apache.org/jira/browse/ORC-104.  I didn't do much research here, but i think it is better to be sync with ORC code. wdyt?

File: common/src/java/org/apache/hadoop/hive/ql/ErrorMsg.java (hunk 3 above)
Comment (zhangbutao) [reply common/src/java/org/apache/hadoop/hive/ql/ErrorMsg.java:527]: I just follow the PARQUET_FOOTER_ERROR, and we can let @jfsii give a explanation. :)

File: ql/src/java/org/apache/hadoop/hive/ql/io/orc/OrcSerde.java (hunk 2 above)
Comment (ayushtkn) [reply ql/src/java/org/apache/hadoop/hive/ql/io/orc/OrcSerde.java:208]: should be fine, was just exploring a possibility. :-)

File: ql/src/java/org/apache/hadoop/hive/ql/io/orc/OrcSerde.java (hunk 1 above)
Comment (zhangbutao) [reply ql/src/java/org/apache/hadoop/hive/ql/io/orc/OrcSerde.java]: Done. Thx.

File: common/src/java/org/apache/hadoop/hive/ql/ErrorMsg.java (hunk 3 above)
Comment (jfsii) [reply common/src/java/org/apache/hadoop/hive/ql/ErrorMsg.java:527]: The source for the change was a comment on a review I had on the patch:
"nit: add a ":" such that the exception is printed after the delimiter"

I could have pushed back on the comment.

**Review by jfsii — APPROVED** [review]
From my end things look good. Thanks again @zhangbutao

**Review by ayushtkn — APPROVED** [review]
LGTM

**Comment by jfsii** [comment]
Is this committable? I don't have commit privs and I don't want to see this patch go stale. cc: @ayushtkn

**Pull request merged**